### ISSUE TITLE
Add UI element for opening PR on github.com

### ISSUE THREAD

**Issue by kuychaco** [desktop/desktop#8152]
So I just pushed and had the desire to view my PR on dotcom. I saw the “View on GitHub” button and I’m like _oh great!_ … And then realized/remembered that this just takes me to the repo. 

I couldn’t find a way via the UI to open my PR page. Eventually I found the “Show Pull Request” menu item and cmd-R shortcut and experienced the sweet relief and satisfaction of doing the thing I wanted to do. And now I’m wondering if it’s worth investing in making this easier for folks…

@tierninho had the great suggestion that we replace the "Create Pull Request" button with an "View Pull Request" button after the PR has been created:

![Screen Shot 2019-08-22 at 1 58 48 PM](https://user-images.githubusercontent.com/7910250/63560265-3b7e7980-c50a-11e9-8e28-d2b3d57bfb6d.png)

**Comment by billygriffin** [comment]
Thanks @kuychaco (and @tierninho for a potential solution!). I agree that it might be useful. The potential problem here is that it gets into the question of hierarchy. For example, when you have commits to push to the branch, do you show the "View this PR" or the "Push these 6 commits" as the prominent suggestion. We delved a bit into that with stashed changes, where we decided that if you have a stash on your branch, "View stash" is always the thing presented as the prominent suggestion. 

But with each addition of a new potentially competing thing, whatever hierarchy we choose presents increasing likelihood of confusion from users who are like, "wait where's the thing I expected?" I'd love to hear some thoughts on how we might address that confusion.

**Comment by billygriffin** [comment]
I'm going to roll https://github.com/desktop/desktop/issues/4114 into this as well and close that one. It's another possible solution for viewing the *current* pull request for a branch from the PR indicator in the branch dropdown.

<img width="236" alt="Screen Shot 2019-11-21 at 8 03 43 AM" src="https://user-images.githubusercontent.com/5091167/69349704-74c3a400-0c35-11ea-91da-2f981486e4c3.png">

**Comment by outofambit** [comment]
fwiw this is should be straightforward to change once we figure out the desired behavior. i'm 👍 to "View Pull Request" as a possible solution, though i agree something persistent might be better.

another option would be replacing this suggested action with a "View PR" one, when a PR is open for this branch. (its always at the bottom of the suggested actions list)

<img width="638" alt="Screen Shot 2019-11-21 at 10 50 54 AM" src="https://user-images.githubusercontent.com/964912/69367305-e0fdd200-0c4c-11ea-9ee2-23e9f1fed28a.png">

**Comment by billygriffin** [comment]
@outofambit That was my suggestion when we discussed this initially if you're on a PR branch, but since it's currently one of the three persistent suggestions it's possible that it might be confusing that it's "View on GitHub" sometimes and "View this PR on GitHub" other times. But what you proposed (replacing the "View on GitHub with "View this PR on GitHub" or something similar) is still my first choice and I think it reasonably reflects a similar intent as the "View on GitHub" space.

**Comment by davidbarton** [comment]
Any progress on this one?

**Comment by 0xdevalias** [comment]
RE: https://github.com/desktop/desktop/issues/8152#issuecomment-557124494

> I'm going to roll #4114 into this as well and close that one. It's another possible solution for viewing the current pull request for a branch from the PR indicator in the branch dropdown.

> Glad to see that v1.1 displays PR number. However, it will be a lot more convenient if that badge is clickable, like in the old client.

>  I'm not sure if this came up earlier (I'll dig that up later) but there's also a keyboard shortcut for this (Cmd+R on macOS or Ctrl+R on Windows) to speed it up.

> Here's some context from the initial exploration [#2642 (comment)](https://github.com/desktop/desktop/issues/2642#issuecomment-330313559)
> 
> > > What should happen when someone clicks on the PR badge? What about the status icon?
> > 
> > 
> > For now, I don't think the badge should have a function. I've been going back and forth on this because it's certainly not far-fetched to imagine it would open the current PR but I'm super vary of overloading the button as well. Buttons in buttons rarely come out well. I have been exploring UI affordances for opening the current PR both in the button and in the foldout but I'm not happy with any of them at the moment and, as @j-f1 points our we will have a way (albeit not a very discoverable one) to get there.
> > TL;DR: Initially nothing, but we'll keep exploring ways of getting quick access to the current PR.

> > is there a possibility of adding the option to right click on a PR in the Current Branch dropdown to provide the Show Pull Request option?
> 
> I'm all for it. I agree with you that it suffers from discoverability problems but it's a stop-gap until we get something better going.

> Tagging this as help wanted for someone who wants to experiment with making a context menu for the branch button which has a "Show Pull Request" action, to see if this makes it easier to navigate to the pull request without going down the path of a complex tooltip.

Based on the above, it seems like a 'Show Pull Request' item to the right-click menu of the 'Branch Switcher' is desirable:

![image](https://github.com/desktop/desktop/assets/753891/d7a6f2bf-7dfd-4631-a973-72c1f44b6cb8)

Inside the expandable is a deep dive into the codebase, and the required functionality that would allow us to implement this, following existing patterns.

<details>
<summary>DeepDive into the Code</summary>

From a quick search in the code, this context menu seems to be generated within `generateBranchContextMenuItems`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/branches/branch-list-item-context-menu.tsx#L11-L40

And could be as simple as adding code such as the following (where the action for `???` still needs to be figured out):

```js
items.push({
  label: 'Show Pull Request',
  action: () => ???,
})
```

To implement the `???` action, it looks like `IBranchContextMenuConfig` would need to be updated, and whatever supporting args needed passed in to `generateBranchContextMenuItems`'s `config` arg:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/branches/branch-list-item-context-menu.tsx#L4-L9

`generateBranchContextMenuItems` seems to be called in 2 places:

`onBranchContextMenu` in `BranchList`: 

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/branches/branch-list.tsx#L269-L290

`onBranchToolbarButtonContextMenu` in `BranchDropdown`:

![image](https://github.com/desktop/desktop/assets/753891/2cd2518c-52aa-4010-97d0-940a0fb85898)

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/toolbar/branch-dropdown.tsx#L230-L249

`BranchList` seems to be used as part of `BranchSelect`, which I don't think we are super interested in here (I didn't fully trace where this actually appears in the app):

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/branches/branch-select.tsx#L97-L118

The `PullRequestBadge` seems to only be rendered as part of `BranchDropdown`, when `this.props.currentPullRequest` (of type `PullRequest`) isn't `null`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/toolbar/branch-dropdown.tsx#L53-L54

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/toolbar/branch-dropdown.tsx#L369-L386

So we probably want to match the current conditions and only add the context menu item where/when the `PullRequestBadge` would show.

The `PullRequest` class is defined here, and contains the PR `number`, as well as the `head` and `base` (both of which are of type `PullRequestRef`):

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/models/pull-request.ts#L22-L32

The `PullRequestRef` class is defined here, and contains things including the `sha` of the commit, and `gitHubRepository` (of type `GitHubRepository`):

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/models/pull-request.ts#L8-L20

The `GitHubRepository` class is defined here:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/models/github-repository.ts#L7-L31

--

From my deep dive on another issue ([Ref](https://github.com/desktop/desktop/issues/9791#issuecomment-1859049098)), I found that there are URL creation helpers like `createCommitURL`, which if we wanted, we could use to add a 'View commit on GitHub' link to this menu:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/commit-url.ts#L4-L24

In GitHub Desktop's 'Branch' menu there is a 'View Pull Request on GitHub' menu item, which is built in `buildDefaultMenu`, and appears to have an id of `create-pull-request`, and call `emit('open-pull-request')` to trigger the functionality:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/main-process/menu/build-default-menu.ts#L459-L464

`emit` is defined here, and seems to trigger the menu with `ipcWebContents.send(window.webContents, 'menu-event', name)`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/main-process/menu/build-default-menu.ts#L691-L707

Searching the codebase for `'menu-event'`, we find the `sendMenuEvent` method within `AppWindow` that seems to also trigger this functionality:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/main-process/app-window.ts#L299-L304

`AppWindow` seems to be used as the `mainWindow`

https://github.com/desktop/desktop/blob/development/app/src/main-process/main.ts#L53

We can see all of the `MenuIDs` here, where we can see `create-pull-request` is defined:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/models/menu-ids.ts#L2-L40

--

Another bit of GitHub Desktop functionality that triggers menus similar to `emit` is `MenuBackedSuggestedAction`, which seems to use `executeMenuItemById` to trigger functionality:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/suggested-actions/menu-backed-suggested-action.tsx#L64-L103

`executeMenuItemById` is defined here, and seems to only take an `id`; so we need to figure out how to get the right `id` for the menu we want to trigger:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/main-process-proxy.ts#L76-L81

Looking at an example of how `MenuBackedSuggestedAction` is used, it seems to use the `getMenuItemInfo` helper to convert a string like `toggle-stashed-changes` into the `menuItemId` used by `executeMenuItemById`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/changes/no-changes.tsx#L427-L446

`getMenuItemInfo` is defined here, which calls `this.getMenuInfoMap(this.props.appMenu)`, which calls `buildMenuItemInfoMap(menu)`, which enumerates the `menu.items` and builds a `Map` of `IMenuItemInfo`s keyed by the `item.id` (recursively handling `submenuItem`s):

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/changes/no-changes.tsx#L206-L208

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/changes/no-changes.tsx#L187-L191

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/changes/no-changes.tsx#L154-L180

I'm not sure if there are other patterns elsewhere in the app for triggering menu items like this; but it might make sense to extract these helper functions out into a common utils/helper file so that it can be more easily reused elsewhere (including in the changes we want to make here)

--

The `appMenu` used by `getMenuInfoMap` is passed into `NoChanges` as a prop:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/changes/no-changes.tsx#L63-L66

`NoChanges` is rendered by `RepositoryView`, which drills the `appMenu` from it's props:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/repository.tsx#L85-L88

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/repository.tsx#L497-L512

`RepositoryView` is rendered from `App`, and sets the `appMenu` prop from `state.appMenuState[0]`, where state is defined as `const state = this.state`: 

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/app.tsx#L3267-L3312

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/app.tsx#L3239

--

`BranchDropdown` is rendered as part of `renderBranchToolbarButton` in `App`, which means it should also be easily able to get access to the `appMenu` from `state.appMenuState[0]`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/app.tsx#L3117-L3156

Even though the `repositoryState` prop is set to `selection.state`, where `selection` is defined above as `this.state.selectedState`, we need access to `state.appMenuState[0]`:

https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/ui/app.tsx#L3118C5-L3118C47

We could start by creating a new helper function within `App` that extracts all of the current usages of `state.appMenuState[0]` to something like `getAppMenuState`.

We could then add a new `appMenu` prop to `BranchDropdown` and pass in the value from `getAppMenuState`.

`BranchDropdown` could then use this new `appMenu` prop, with the `getMenuInfoMap` helpers we extracted from `NoChanges`, along with `executeMenuItemById`, to create a helper function that will trigger the `create-pull-request` menu ID used by the `Branch -> View Pull Request on GitHub` menu item, which can be passed into `generateBranchContextMenuItems`, and finally used to implement the `???` action of our new menu item.

</details>

---

RE: https://github.com/desktop/desktop/issues/8152#issuecomment-557223951

> another option would be replacing this suggested action with a "View PR" one, when a PR is open for this branch. (its always at the bottom of the suggested actions list)

RE: https://github.com/desktop/desktop/issues/8152#issuecomment-557229958

> But what you proposed (replacing the "View on GitHub with "View this PR on GitHub" or something similar) is still my first choice and I think it reasonably reflects a similar intent as the "View on GitHub" space.

Another potential here (though I don't believe the `MenuBackedSuggestedAction` / `SuggestedAction` / etc components currently support it, at least based on my deep dive in another issue ([Ref](https://github.com/desktop/desktop/issues/12864#issuecomment-1859062328))), could be to show the primary button in this section as 'View on GitHub', and then if a PR for this branch exists, show a secondary button of 'View this PR on GitHub' (or switch the primary/secondary around)

It might be considered to be complicating the UI in this section too much; but the ability to have a primary/secondary button would also be useful for suggestions like this one:

- https://github.com/desktop/desktop/issues/12864
